### PR TITLE
fix: overwrite ubuntu1604 distro for upgrade and block deploy and generate scenarios on Azure Stack Hub

### PR DESCRIPTION
**Pull request by haofan-ms**
**Reason for Change**:

Fix the previously documented manual step, update ubuntu distro information for cluster created with "aks-ubuntu-16.04" during deploy, generate and upgrade operations.

**Issue Fixed**:

**Credit Where Due**:

Does this change contain code from or inspired by another project?

- [ ] No
- [ ] Yes

If "Yes," did you notify that project's maintainers and provide attribution?

- [ ] No
- [ ] Yes

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

File: cmd/upgrade.go [hunk 1]
```diff
@@ -207,6 +207,21 @@ func (uc *upgradeCmd) loadCluster() error {
 		}
 	}
 
+	// Update the masterProfile and agentPoolProfiles distro for AzureStackCloud to use aks-ubuntu-18.04 instead of aks-ubuntu-16.04
+	if uc.containerService.Properties.IsAzureStackCloud() {
+		if uc.containerService.Properties.MasterProfile.Distro == api.AKSUbuntu1604 {
+			log.Infoln("aks-ubuntu-16.04 support has depricated, updating the master distro to aks-ubuntu-18.04")
```
Comment (JesusAlvarezTorres) [inline cmd/upgrade.go:213]: typo, it is deprecated not depricated

File: cmd/upgrade.go [hunk 2]
```diff
@@ -207,6 +207,21 @@ func (uc *upgradeCmd) loadCluster() error {
...
+			uc.containerService.Properties.MasterProfile.Distro = api.AKSUbuntu1804
+		}
+
+		for _, app := range uc.containerService.Properties.AgentPoolProfiles {
+			if app.Distro == api.AKSUbuntu1604 {
+				log.Infoln(fmt.Sprintf("aks-ubuntu-16.04 support has depricated, updating the %s distro to aks-ubuntu-18.04", app.Name))
```
Comment (JesusAlvarezTorres) [inline cmd/upgrade.go:219]: same typo as in the master profile

File: cmd/upgrade.go (hunk 1 above)
Comment (JesusAlvarezTorres) [inline cmd/upgrade.go:213]: typo, it is deprecated not depricated

File: cmd/upgrade.go (hunk 2 above)
Comment (haofan-ms) [reply cmd/upgrade.go]: Updated

File: cmd/upgrade.go (hunk 1 above)
Comment (haofan-ms) [reply cmd/upgrade.go]: Updated

**Review by JesusAlvarezTorres — APPROVED** [review]
lgtm

File: cmd/deploy.go [hunk 3]
```diff
@@ -238,6 +238,20 @@ func (dc *deployCmd) loadAPIModel() error {
 				dc.containerService.Properties.CustomCloudProfile.IdentitySystem = dc.authProvider.getAuthArgs().IdentitySystem
 			}
 		}
+
+		if dc.containerService.Properties.IsAzureStackCloud() {
+			if dc.containerService.Properties.MasterProfile.Distro == api.AKSUbuntu1604 {
+				log.Errorln("Distro 'aks-ubuntu-16.04' is not longer supported on Azure Stack Hub, please update master profile distro in api model to use 'aks-ubuntu-18.04'")
```
Comment (jadarsie) [inline cmd/deploy.go:244]: I think the result of these two lines will be two similar messages in stdout. If that's the case, can you just keep one?

File: cmd/deploy.go [hunk 4]
```diff
@@ -238,6 +238,20 @@ func (dc *deployCmd) loadAPIModel() error {
...
+				return errors.New("Distro value 'aks-ubuntu-16.04' of master profile is not supported on Azure Stack Hub")
+			}
+
+			for _, app := range dc.containerService.Properties.AgentPoolProfiles {
```
Comment (jadarsie) [inline cmd/deploy.go:248]: Could you please add unit tests to avoid regressions?

File: cmd/upgrade.go [hunk 5]
```diff
@@ -207,6 +207,21 @@ func (uc *upgradeCmd) loadCluster() error {
...
+	// Update the masterProfile and agentPoolProfiles distro for AzureStackCloud to use aks-ubuntu-18.04 instead of aks-ubuntu-16.04
+	if uc.containerService.Properties.IsAzureStackCloud() {
+		if uc.containerService.Properties.MasterProfile.Distro == api.AKSUbuntu1604 {
+			log.Infoln("Distro 'aks-ubuntu-16.04' is not longer supported on Azure Stack Hub, overwriting master profile distro to 'aks-ubuntu-18.04'")
+			uc.containerService.Properties.MasterProfile.Distro = api.AKSUbuntu1804
+		}
+
+		for _, app := range uc.containerService.Properties.AgentPoolProfiles {
+			if app.Distro == api.AKSUbuntu1604 {
+				log.Infoln(fmt.Sprintf("Distro 'aks-ubuntu-16.04' is not longer supported on Azure Stack Hub, overwriting agent profile %s distro to 'aks-ubuntu-18.04'", app.Name))
```
Comment (jadarsie) [inline cmd/upgrade.go:219]: agentpool profile

File: docs/topics/azure-stack.md [hunk 6]
```diff
@@ -300,7 +300,9 @@ There is no official support for private-preview Kubernetes cluster with Windows
 
 ### Upgrading Kubernetes clusters created with the Ubuntu 16.04 distro
 
-Starting with AKS Engine v0.63.0, the Ubuntu 16.04 distro is not longer a supported option as the OS reached its end-of-life. In order to upgrade a cluster, make sure to set the OS distro to `aks-ubuntu-18.04` in your input API model, the one generated by `aks-engine deploy` and passed as input to `aks-engine upgrade`.
+Starting with AKS Engine v0.63.0, the Ubuntu 16.04 distro is not longer a supported option as the OS reached its end-of-life. For kubernetes clusters created with `aks-ubuntu-16.04` distro, it will automatically get upgrade to `aks-ubuntu-18.04`, corresponding distro values in `masterProfile` and `agentPoolProfile`s of the API model will be updated as well.
```
Comment (jadarsie) [inline docs/topics/azure-stack.md:303]: I think that this text gives the idea that v0.63.0 will overwrite the distro value, which is not true.

File: cmd/deploy.go (hunk 4 above)
Comment (haofan-ms) [reply cmd/deploy.go:248]: Added

**Review by jadarsie — APPROVED** [review]

**Pull request merged**